torch/_dynamo/variables/torch.py

```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
             return handle_ntuple(args[0])
 
 
+def is_fn_safe_to_run(tx, f, sub_args):
+    # Snapshot state
+    graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
+    # Will raise if not sound
+    output, graph, lifted_freevars = speculate_subgraph(
+        tx, f, sub_args, graph_checkpoint, checkpoint, always_restore=True
+    )
+    # If we got here, we are probably sound, but we do not support freevars yet, so lets call cases where we found them
+    # unsafe.
+    if lifted_freevars:
+        return False
+    return True
+
+
+# See NOTE [HigherOrderOperator tracing design] for details of the design
+# See NOTE [speculate_subgraph vs old_speculate_subgraph] for other info
+def speculate_subgraph(
+    tx, f, sub_args, graph_checkpoint, checkpoint, *, always_restore=False
+):
+    from . import AutogradFunctionContextVariable, ConstantVariable, TensorVariable
+
+    try:
+        with tx.output.new_subtracer() as tracer:
+            args = []
+            # One argument to graph per sub_args
+            for a in sub_args:
+                assert not isinstance(
+                    a, torch.Tensor
+                ), "Tensors should already be tracked?"
+                if isinstance(a, ConstantVariable):
+                    proxy = tracer.create_graph_input("const")
+                elif isinstance(a, (TensorVariable, AutogradFunctionContextVariable)):
+                    tracer.create_graph_input(a.as_proxy().node.name)
```

Comment:
This makes me a bit uncomfortable, but I think I can live with it:
- let's say we access ctx.saved_tensors[0]. Will this trigger our free variable check?
- can anything be a proxy, or are proxies supposed to be Tensors?

One way around this is that we can have trampoline_fwd and trampoline_bwd "hide" the ctx object. That is, trampoline_fwd could instead return (output, saved_tensors) and trampoline_bwd could accept (actual_args, saved_tensors), and then inside trampoline_bwd we construct a new ctx object and install saved_tensors on it.

Reply:
Hmm, I don't hate that idea, but I worry that this might lead to us having to model more than saved_tensors, whereas now AutogradFunctionContextVariable wrapping a real object does a lot of heavy lifting for us.

Reply:
Fixed by lazy